server/lib/cacheProvider.js

```diff
@@ -0,0 +1,130 @@
+import LRU from 'lru-cache';
+import memjs from 'memjs';
+import redis from 'redis';
+import { has, get, isFunction } from 'lodash';
+import { promisify } from 'util';
+
+import logger from './logger';
+
+const debugCache = debug('cache');
+
+export const PROVIDER_TYPES = {
+  MEMCACHE: 'MEMCACHE',
+  MEMORY: 'MEMORY',
+  REDIS: 'REDIS',
+};
+
+let defaultProvider = PROVIDER_TYPES.MEMORY;
+if (has(config, 'memcache.servers')) defaultProvider = PROVIDER_TYPES.MEMCACHE;
+else if (has(config, 'redis.serverUrl')) defaultProvider = PROVIDER_TYPES.REDIS;
+
+const makeMemcacheProvider = ({ servers, username, password }, { serialize, unserialize }) => {
+  const client = memjs.Client.create(servers, { username, password });
+  return {
+    clear: async () => client.flush(),
+    del: async key => client.delete(key),
+    get: async key => {
+      const data = await client.get(key);
+      if (data.value) {
+        const value = data.value.toString();
+        try {
+          return unserialize(value);
+        } catch (err) {
+          debugCache(`Invalid JSON (${value}): ${err}`);
+        }
+      }
+    },
+    has: async key => {
+      const value = await client.get(key);
+      return value !== undefined;
+    },
+    set: async (key, value, expirationInSeconds = 0) => {
+      if (value !== undefined) return client.set(key, serialize(value), { expires: expirationInSeconds });
+    },
+  };
+};
+
+const makeRedisProvider = ({ serverUrl, opts }, { serialize, unserialize }) => {
+  const client = redis.createClient(serverUrl, opts);
+  const asyncFlushAll = promisify(client.flushall).bind(client);
+  const asyncDel = promisify(client.del).bind(client);
+  const asyncGet = promisify(client.get).bind(client);
+  const asyncSet = promisify(client.set).bind(client);
+  const asyncSetex = promisify(client.setex).bind(client);
+  return {
+    clear: async () => asyncFlushAll(),
+    del: async key => asyncDel(key),
+    get: async key => {
+      const value = await asyncGet(key);
+      if (value) {
+        try {
+          return unserialize(value);
+        } catch (err) {
+          debugCache(`Invalid JSON (${value}): ${err}`);
+        }
+      } else return undefined;
+    },
+    has: async key => {
+      const value = await asyncGet(key);
+      return value !== null;
+    },
+    set: async (key, value, expirationInSeconds) => {
+      if (value !== undefined && expirationInSeconds) return asyncSetex(key, expirationInSeconds, serialize(value));
+      if (value !== undefined) return asyncSet(key, serialize(value));
+    },
+  };
+};
+
+const makeMemoryProvider = opts => {
+  const lruCache = new LRU(opts);
+  return {
+    clear: async () => lruCache.reset(),
+    del: async key => lruCache.del(key),
+    get: async key => lruCache.get(key),
+    has: async key => lruCache.has(key),
+    set: async (key, value, expirationInSeconds) => lruCache.set(key, value, expirationInSeconds * 1000),
+  };
+};
+
+const unfailable = provider =>
```

Comment:
While we get it, we usually prefer simpler code.

Reply:
Yeah, I agree 100%.
Proxy is the kind of thing you should only use in libraries for very specific reasons.